photoeditor/src/main/java/ja/burhanrashid52/photoeditor/DrawingView.kt

```diff
@@ -54,8 +54,8 @@ class DrawingView @JvmOverloads constructor(
         // apply shape builder parameters
         currentShapeBuilder?.apply {
             paint.strokeWidth = this.shapeSize
-            paint.alpha = this.shapeOpacity
             paint.color = this.shapeColor
```

Comment:
Alpha (opacity) for `Paint` object can be set in two ways. Using `Paint.setAlpha()` or `Paint.setColor()`. Color in Android is AARRGGBB (for example: 0xFF0000FF - blue with no transparency). Calling `setAlpha()` modifies AA value of currently set color. That means if you call `setAlpha()` followed by `setColor()`, this will set completely new color with new AA value. That's the reason why the order of method calls matter. And that's the reason why setting `shapeOpacity` is currently broken. If you call `setColor()` followed by `setAlpha()`, alpha value set in `setColor()` will be overwritten by `setAlpha()`. `shapeOpacity` has default value `0`. That means that if we set in the app's code `shapeColor` and we don't set `shapeOpacity`, our AA value from `shapeColor` will be overwritten by default value of `shapeOpacity` which is `0`. This will break setting opacity using AA value from `shapeColor`.

My suggestion is to make `shapeOpacity` nullable with default value `null` and not to call `Paint.setAlpha()` when `shapeOpacity` is `null`. This solution overrides AA from `setColor()` only if `shapeOpacity` is explicitly set.

In addition to your changes I suggest:

- In `DrawingView.kt:58` replace `paint.alpha = this.shapeOpacity` with `shapeOpacity?.also { paint.alpha = it }`

- In `ShapeBuilder:19` replace `var shapeOpacity = 0` with `var shapeOpacity: Int? = null`

- To prevent mistakes in the future I think it's good idea to add a comment in the code in the line before `paint.color` and `paint.alpha` calls (`DrawingView.kt:57`):

```
// 'paint.color' must be called before 'paint.alpha',
// otherwise 'paint.alpha' value will be overwritten.
```

Reply:
ok!

Reply:
It's done my friend, and thanks for the feedback! If you need anything else, I'm available.

Reply:
Thank you for your updates. I missed two more changes in my previous comment. I apologize for that.

- `ShapeBuilder:41`: `opacity: Int` should be changed to `opacity: Int?`
- `ShapeBuilder:54`: `const val DEFAULT_SHAPE_OPACITY = 255` should be changed to `val DEFAULT_SHAPE_OPACITY = null`

Reply:
That's what we're here for, I'm grateful and happy to be able to contribute and learn, with this magical code! And thanks again for the feedback.

Reply:
Wow!! Nice explaination @wasky. I think you should write a blog post explaining this issue in detail :)

Few things.
@wasky Is there any way to add a test for this? Using Robolectric or some by writing a test against `ShaperBuilder?

Reply:
> Wow!! Nice explanation @wasky. I think you should write a blog post explaining this issue in detail :)

Thanks! :)

> Is there any way to add a test for this? Using Robolectric or some by writing a test against `ShaperBuilder`?

I'm not sure if we can test it using Roboelectric. In Roboelectric tests `setAlpha()` doesn't modify `AA` value from `getColor()`. I don't have experience with Roboelectric, so I don't know why Roboelectic behaviour is different than actual Android code. But we can test it using standard android tests. I can write a test for this as a different PR after this will be merged if it's OK for you?

Reply:
@wasky Works for me. Created #496